src/mo_frontend/bi_match.ml

```diff
@@ -341,38 +341,26 @@ let bi_match_typs ctx =
 let solve ctx (ts1, ts2) deferred_typs =
   if debug then Debug.print_solve ctx (ts1, ts2);
 
-  (* Defer solving type variables that:
-    - appear ONLY in deferred func inputs. There is no info to help us infer them, it's better to raise an error than infer a default bound that could lead to confusing errors.
-    - appear in deferred func outputs (unless they appear in inputs). Solving them now could restrict the solution.
-   *)
+  (* Defer solving type variables that can be solved later. More constraints appear in the next round, let them influence as many variables as possible *)
   let to_defer, defer_verify = if deferred_typs = [] then (ConSet.empty, false) else
     (* Type variables mentioned/used in subtyping constraints *)
     let cons1 = cons_typs ts1 in
     let cons2 = cons_typs ts2 in
     let used = ConSet.inter ctx.var_set (ConSet.union cons1 cons2) in
+    let unused = ConSet.diff ctx.var_set used in
 
-    let inputs, outputs = List.fold_left (fun (inputs, outputs) t ->
+    (* Solve only variables that need to be solved now (inputs of deferred functions must be solved before the next round) *)
+    let to_solve = List.fold_left (fun acc t ->
       match normalize t with
-      | Func (_, _, _, t1, t2) ->
-        let inputs = ConSet.union inputs (cons_typs t1) in
-        let outputs = ConSet.union outputs (cons_typs t2) in
-        (inputs, outputs)
-      | _ ->
-        (ConSet.union inputs (cons t), outputs)
-      ) (ConSet.empty, ConSet.empty) deferred_typs
+      | Func (_, _, _, t1, _) -> ConSet.union acc (cons_typs t1)
```

Comment:
Does it matter that t1 might contain Var's from the Func (it might have non-empty bounds). Does cons_typs expect to see a freestanding Var i ? Might be worth checking.

Reply:
Or maybe you should strengthen the case pattern to
| Func (_, _, [], t1, _) -> ConSet.union acc (cons_typs t1)